numpy/ma/core.py

```diff
@@ -5079,24 +5002,64 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            with np.errstate(divide='ignore', invalid='ignore'):
+                ret = self._data.mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(True if zerocounts else nomask)
```

Comment:
This seems to assume `zerocounts` is always a single-element item, not an array, which would only be the case for `axis=None`. I think this needs to be `__setmask__(zerocounts if zerocounts.any() else nomask)` or something similar.

But... Is it possible for there to be zero counts? Here, you are in the case where there is no mask, so it would happen only if any of the dimensions were 0, correct? Or are we considering `NaN` here? I'm quite confused...